homework05/test/examples/Main.hs

```diff
@@ -1,15 +1,61 @@
-module Main
-  ( main
-  )
-  where
-
--- doctest
-import qualified Test.DocTest as DocTest
+module Main where
 
+import ExprT
+import Calc
+import Parser
+import StackVM
+import Test.Hspec
 
 main :: IO ()
-main =
-  DocTest.doctest
-    [ "-isrc"
-    , "src/Calc.hs"
-    ]
+main = hspec $ do
+  
+  describe "eval" $
+    it "returns result of applying add and multiply operations" $
+      eval (ExprT.Mul (ExprT.Add (ExprT.Lit 2) (ExprT.Lit 3)) (ExprT.Lit 4)) `shouldBe` 20
+
+  describe "evalStr" $ do
+    it "evaluates expression with parenthesis" $
+      evalStr "(2+3)*4" `shouldBe` Just 20
+
+    it "evaluates expression without parenthesis" $
+      evalStr "2+3*4" `shouldBe` Just 14
+
+    it "evaluates wrong expression" $
+      evalStr "2+3*" `shouldBe` Nothing
+      
+  describe "reify" $
+    it "makes an instance of Expr to ExprT type" $
+      reify (mul (add (lit 2) (lit 3)) (lit 4)) `shouldBe` ExprT.Mul (ExprT.Add (ExprT.Lit 2) (ExprT.Lit 3)) (ExprT.Lit 4)
+  
+  describe "parseExp" $
+    context "when provided with multiple data types" $ do
+      it "returns value of the expression evaluated as Integer" $ do
+        let testInteger = parseExp lit add mul "(3 * 4) + (-5)" :: Maybe Integer
+        testInteger `shouldBe` Just 7
+    
+      it "returns value of the expression evaluated as Bool" $ do
+        let testBool = parseExp lit add mul "(3 * 4) + (-5)" :: Maybe Bool
+        testBool `shouldBe` Just True
+    
+      it "returns value of the expression evaluated as MinMax" $ do
+        let testMM = parseExp lit add mul "(3 * 4) + (-5)" :: Maybe MinMax
+        testMM `shouldBe` Just (MinMax 3)
+    
+      it "returns value of the expression evaluated as Mod7" $ do
+        let testSat = parseExp lit add mul "(3 * 4) + (-5)" :: Maybe Mod7
+        testSat `shouldBe` Just (Mod7 0)
+   
+  describe "compile" $
```

Comment:
`parseExp` function in `Parser.hs` file handles this case, producing `Nothing` for the inputs which are not well-formed expressions.